File: monitor/bin/js/block-data-attempts.js
const { mapBy, readLines, shell } = require('ez-tools');
const { debuglog } = require('util');
const debuglogFunctions = debuglog('functions');
const db = require(`${__dirname}/db-production.js`);

/*
  Firewall reject all IPs that have been assessed as data theft attempt
*/

async function blockDataAttempts() {
  debuglogFunctions('blockDataAttempts');
  let sql = 'SELECT ip FROM ip WHERE data_attempt = TRUE';
  let result = await db.query(sql);
  let dataAttemptIps = mapBy(result.rows, 'ip');
  let dataAttemptRanges = dataAttemptIps.map(ip => `${ip}/32`);
  let rejectString = dataAttemptIps.sort().join(',');
  let blockCommand = `gcloud compute --project=${process.env.RESUME_PROJECT_ID} `;
      blockCommand += 'firewall-rules update no-hackers ';
      blockCommand += `--source-ranges=${rejectString}`;
  await shell(blockCommand);
  db.end();
  console.log('cya');
}

blockDataAttempts();

File: monitor/bin/rb/block_data_attempts.rb
#!/usr/bin/env ruby

require_relative 'helpers/db_production'
require_relative 'helpers/get_blocked_hacker_ips'

# Firewall reject all IPs that have been flagged as data attempt

def block_hackers()
  get_hackers_sql = "SELECT ip FROM ip WHERE hacker = TRUE AND ip != '#{ENV['MY_IP']}'";
  db.exec(get_hackers_sql) do |result|
    hacker_ips = result.map { |row| row['ip'] }
    blocked_hacker_ips = get_blocked_hacker_ips()
    all_ips_to_block = hacker_ips.concat(blocked_hacker_ips).uniq()
    ranges_to_block = all_ips_to_block.map { |ip| "#{ip}/32" }
    reject_string = ranges_to_block.sort().join(',')
    block_command = "gcloud compute --project=#{ENV['RESUME_PROJECT_ID']} "
    block_command += 'firewall-rules update no-hackers '
    block_command += "--source-ranges=#{reject_string}"
    puts block_command
    result = `#{block_command}`
    print result
    describe_command = "gcloud compute --project=#{ENV['RESUME_PROJECT_ID']} "
    describe_command += 'firewall-rules describe no-hackers'
    puts describe_command
    result = `#{describe_command}`
    puts result
    puts 'cya'
  end
end

block_hackers()
